File: tools/studio_tracks.py
#!/usr/bin/env python3
"""What a track is, on disk.

Split out of studio.py at the 500-line cap, along the seam that was already
there: none of this knows about HTTP. It answers "which files are tracks",
"where does this id live", and "what does the panel need to know about it" —
and the server above it does the routing.

The id is the contract everywhere else in the project: scenes reference it,
the manifest keys on it, the panel displays it. Which container it happens to
live in is an import detail that stops here.
"""

from __future__ import annotations

import os
import re
from pathlib import Path

import analyze as ana
import manifest as mf

ROOT = Path(__file__).resolve().parent.parent
# Overridable so a test can drive the real server against a disposable
# directory instead of the tracks you actually care about. Nothing else reads
# it; leave it unset and this is the repo's own tracks/.
TRACKS = Path(os.environ.get("CASTLE_TRACKS") or (ROOT / "tracks"))

# Every container import_track.py can write. This list is the reason the format
# option works at all end to end: globbing "*.mp3" — which is what this file did
# when only MP3 existed — makes a WAV or FLAC import land on disk and then never
# appear in the panel, which reads as the import having silently failed.
AUDIO_EXT = ("mp3", "wav", "flac", "opus")
MIME = {
    "mp3": "audio/mpeg",
    "wav": "audio/wav",
    "flac": "audio/flac",
    "opus": "audio/ogg",
}


def track_files() -> list[Path]:
    """Every imported track, whatever container it landed in, by id."""
    return sorted(
        (p for e in AUDIO_EXT for p in TRACKS.glob(f"*.{e}")), key=lambda p: p.stem
    )


def parse_sensitivity(q: dict) -> float | dict:
    """Read `?sensitivity=` plus any per-band `?sens_low=` overrides.

    The three bands routinely want different thresholds — a track can have a
    crisp kick under a wash of cymbals — so the editor sends one per band. A
    bare `sensitivity` still means "all three", which is what every older
    caller sends.
    """
    base = 1.1
    try:
        base = float((q.get("sensitivity") or ["1.1"])[0])
    except ValueError:
        pass
    per = {}
    for short in ("low", "mid", "high"):
        raw = (q.get(f"sens_{short}") or [None])[0]
        if raw is None:
            continue
        try:
            per[f"onset_{short}"] = float(raw)
        except ValueError:
            continue
    if not per:
        return base
    # Any band the caller did not name keeps the shared value.
    for short in ("low", "mid", "high"):
        per.setdefault(f"onset_{short}", base)
    return per


SRC_DIR = "_src"  # kept originals of dropped/pulled files, beside the library


def source_copies(tid: str) -> list[Path]:
    """The kept original(s) for a track — what Delete must take with it."""
    return sorted((TRACKS / SRC_DIR).glob(f"{tid}.*"))


#: What an id may contain, matching what import_track will WRITE (letters,
#: digits, underscore — see its explicit-id guard). Spelled here as well so
#: the read path is guarded at its own choke point rather than relying on
#: every caller to have stripped the name first: an id becomes a filename,
#: and a filename with a separator or a parent hop in it is not an id.
ID_RE = re.compile(r"^\w{1,64}$", re.ASCII)


def valid_id(tid: str) -> bool:
    """Is `tid` a track id, rather than something wearing one's clothes?"""
    return bool(ID_RE.match(tid))


def track_path(tid: str) -> Path | None:
    """Resolve a bare track id to the file that holds it.

    The id is the contract everywhere else — scenes, the manifest, the panel —
    and the extension is an import detail. Callers pass the id and get back
    whichever container it happens to live in, or None. An id that is not one
    resolves to nothing at all, whatever it was hoping to reach.
    """
    if not valid_id(tid):
        return None
    for e in AUDIO_EXT:
        p = TRACKS / f"{tid}.{e}"
        if p.exists():
            return p
    return None


def source_missing(source: str) -> bool:
    """True for a file: source whose file is no longer there."""
    return source.startswith("file:") and not Path(source[len("file:") :]).exists()


def track_infos(paths: list[Path]) -> list[dict]:
    """track_info for a whole listing, reading tracks.json ONCE.

    /api/tracks used to load and parse the manifest once per track; that is
    nothing at two tracks and a file read per row at twenty. The route
    should call this rather than mapping track_info over the listing."""
    data = mf.load()
    return [track_info(p, data.get(p.stem) or {}) for p in paths]


def track_info(p: Path, meta: dict | None = None) -> dict:
    """Everything the Tracks panel needs, including where the file came from.

    The manifest is the cheap part — read it even if decoding fails, so a
    broken file still shows its source and can be re-imported. `meta` is
    this track's manifest entry when the caller already holds the whole
    file (track_infos); left None, it is looked up here as before.
    """
    if meta is None:
        meta = mf.get(p.stem) or {}
    info = {
        "id": p.stem,
        # The panel needs this to write `audio_file:` into a scene. Without it
        # it guessed ".mp3", which produced a scene pointing at a file that
        # does not exist for every non-MP3 import.
        "ext": p.suffix.lstrip("."),
        "kb": p.stat().st_size // 1024,
        # Exact size, so the desk can tell a CURRENT card copy from a STALE
        # one (re-imported since it was sent) by comparing against /api/files.
        "bytes": p.stat().st_size,
        "source": meta.get("source", ""),
        # A dropped file's original lives in tracks/_src/ (import_track
        # --keep-source); one that was imported from a path that has since
        # gone cannot be re-imported, and the panel must say so instead of
        # offering a button that fails with an absolute path (JB1-3).
        "source_missing": source_missing(meta.get("source", "")),
        "title": meta.get("title", ""),
        "imported": meta.get("imported", ""),
        "opts": meta.get("opts", {}),
        "notes": meta.get("notes", ""),
    }
    cached = _from_manifest(meta, p.stat().st_size)
    if cached is not None:
        info.update(cached)
        return info
    try:
        x = ana.load_audio(p)
        marks = ana.analyze(x)
    except Exception as e:
        info["error"] = str(e)
        return info
    info["dur"] = round(len(x) / ana.SR, 2)
    info["onsets"] = {k: len(v) for k, v in marks.items()}
    # Remember the answer beside the provenance, so the next /api/tracks
    # reads it instead of decoding the library again (it was linear in the
    # number of tracks, every call). `bytes` is the staleness check.
    mf.patch(
        p.stem,
        audio={
            **meta.get("audio", {}),
            "duration": info["dur"],
            "bytes": info["bytes"],
        },
        onsets=info["onsets"],
    )
    return info


def _from_manifest(meta: dict, size: int) -> dict | None:
    """The decode-free answer, if the manifest has one for THIS file.

    import_track.py records duration, byte size and per-band onset counts
    at import time; the size doubles as the staleness check, so a file
    replaced out of band (same id, different bytes) is decoded afresh. The
    import's level_* entries (envelope points for beatless bands) are not
    onsets and are left out, matching what a live analysis reports.
    """
    audio = meta.get("audio") or {}
    onsets = meta.get("onsets")
    if "duration" not in audio or not isinstance(onsets, dict):
        return None
    if audio.get("bytes") != size:
        return None
    return {
        "dur": round(float(audio["duration"]), 2),
        "onsets": {k: int(v) for k, v in onsets.items() if k.startswith("onset_")},
    }

File: core/src/studio_tracks.rs
//! What a track is, on disk — tools/studio_tracks.py, field for field.
//!
//! The id is the contract everywhere else in the project; which container
//! it lives in is an import detail that stops here. track_info's answers
//! are held JSON-equal to the Python's by tests/test_studio_rust.py — the
//! analysis half (duration, per-band onset counts) rides the crate's own
//! bit-exact media/onsets port, so the numbers agree by construction.

use std::path::{Path, PathBuf};

use crate::jsonio::{Json, obj_update};
use crate::{manifest, media, onsets};

pub const AUDIO_EXT: [&str; 4] = ["mp3", "wav", "flac", "opus"];
const SR: f64 = 44100.0;
pub const SRC_DIR: &str = "_src";

pub fn mime(ext: &str) -> &'static str {
    match ext {
        "mp3" => "audio/mpeg",
        "wav" => "audio/wav",
        "flac" => "audio/flac",
        "opus" => "audio/ogg",
        _ => "application/octet-stream",
    }
}

/// studio_tracks.ID_RE — `^\w{1,64}$`, ASCII.
pub fn valid_id(tid: &str) -> bool {
    !tid.is_empty()
        && tid.len() <= 64
        && tid.bytes().all(|b| b.is_ascii_alphanumeric() || b == b'_')
}

/// Every imported track, whatever container it landed in, by id.
pub fn track_files(tracks: &Path) -> Vec<PathBuf> {
    let mut out: Vec<PathBuf> = Vec::new();
    for ext in AUDIO_EXT {
        let Ok(dir) = std::fs::read_dir(tracks) else {
            continue;
        };
        let mut batch: Vec<PathBuf> = dir
            .flatten()
            .map(|e| e.path())
            .filter(|p| {
                p.is_file()
                    && p.extension().and_then(|e| e.to_str()) == Some(ext)
                    && !p
                        .file_name()
                        .and_then(|n| n.to_str())
                        .unwrap_or("")
                        .starts_with('.')
            })
            .collect();
        batch.sort();
        out.extend(batch);
    }
    out.sort_by_key(|p| stem_of(p));
    out
}

fn stem_of(p: &Path) -> String {
    p.file_stem()
        .and_then(|s| s.to_str())
        .unwrap_or("")
        .to_string()
}

/// Resolve a bare track id to the file that holds it, or None.
pub fn track_path(tracks: &Path, tid: &str) -> Option<PathBuf> {
    if !valid_id(tid) {
        return None;
    }
    AUDIO_EXT
        .iter()
        .map(|e| tracks.join(format!("{tid}.{e}")))
        .find(|p| p.exists())
}

/// The kept original(s) for a track — what Delete must take with it.
pub fn source_copies(tracks: &Path, tid: &str) -> Vec<PathBuf> {
    let Ok(dir) = std::fs::read_dir(tracks.join(SRC_DIR)) else {
        return Vec::new();
    };
    let mut out: Vec<PathBuf> = dir
        .flatten()
        .map(|e| e.path())
        .filter(|p| {
            p.file_name()
                .and_then(|n| n.to_str())
                .is_some_and(|n| n.starts_with(&format!("{tid}.")))
        })
        .collect();
    out.sort();
    out
}

/// True for a `file:` source whose file is no longer there.
pub fn source_missing(source: &str) -> bool {
    source
        .strip_prefix("file:")
        .is_some_and(|p| !Path::new(p).exists())
}

/// track_info for a whole listing, reading tracks.json once.
pub fn track_infos(tracks: &Path) -> Vec<Json> {
    let data = manifest::load(tracks);
    track_files(tracks)
        .iter()
        .map(|p| {
            let meta = data
                .iter()
                .find(|(k, _)| *k == stem_of(p))
                .map(|(_, v)| v.clone())
                .unwrap_or_else(Json::obj);
            track_info(p, &meta, tracks)
        })
        .collect()
}

/// Everything the Tracks panel needs, including where the file came from.
pub fn track_info(p: &Path, meta: &Json, tracks: &Path) -> Json {
    let stem = stem_of(p);
    let size = std::fs::metadata(p).map(|m| m.len()).unwrap_or(0) as i64;
    let source = meta.str_or("source", "");
    let mut info: Vec<(String, Json)> = vec![
        ("id".into(), Json::Str(stem.clone())),
        (
            "ext".into(),
            Json::Str(
                p.extension()
                    .and_then(|e| e.to_str())
                    .unwrap_or("")
                    .to_string(),
            ),
        ),
        ("kb".into(), Json::Int(size / 1024)),
        ("bytes".into(), Json::Int(size)),
        ("source".into(), Json::Str(source.clone())),
        ("source_missing".into(), Json::Bool(source_missing(&source))),
        ("title".into(), Json::Str(meta.str_or("title", ""))),
        ("imported".into(), Json::Str(meta.str_or("imported", ""))),
        (
            "opts".into(),
            meta.get("opts").cloned().unwrap_or_else(Json::obj),
        ),
        ("notes".into(), Json::Str(meta.str_or("notes", ""))),
    ];
    if let Some(cached) = from_manifest(meta, size) {
        info.extend(cached);
        return Json::Obj(info);
    }
    let Some(x) = p.to_str().and_then(media::load_audio) else {
        // The Python surfaces load_audio's exception text; the shape (an
        // `error` key beside the base fields) is the contract.
        info.push((
            "error".into(),
            Json::Str(format!("could not decode {}", p.display())),
        ));
        return Json::Obj(info);
    };
    let dur = onsets::round2(x.len() as f64 / SR);
    let marks = onsets::analyze(&x, 1.1);
    info.push(("dur".into(), Json::Num(dur)));
    let counts: Vec<(String, Json)> = marks
        .iter()
        .map(|(k, v)| (k.clone(), Json::Int(v.len() as i64)))
        .collect();
    info.push(("onsets".into(), Json::Obj(counts.clone())));
    // Remember the answer beside the provenance — the next listing reads it
    // instead of decoding the library again; `bytes` is the staleness check.
    let mut audio: Vec<(String, Json)> = meta
        .get("audio")
        .and_then(Json::as_obj)
        .map(<[(String, Json)]>::to_vec)
        .unwrap_or_default();
    obj_update(
        &mut audio,
        vec![
            ("duration".into(), Json::Num(dur)),
            ("bytes".into(), Json::Int(size)),
        ],
    );
    let _ = manifest::patch(
        tracks,
        &stem,
        vec![
            ("audio".into(), Json::Obj(audio)),
            ("onsets".into(), Json::Obj(counts)),
        ],
    );
    Json::Obj(info)
}

/// The decode-free answer, if the manifest has one for THIS file.
fn from_manifest(meta: &Json, size: i64) -> Option<Vec<(String, Json)>> {
    let audio = meta.get("audio")?;
    let duration = audio.get("duration")?.as_f64()?;
    let onsets_v = meta.get("onsets")?.as_obj()?;
    if audio.get("bytes")?.as_f64()? != size as f64 {
        return None;
    }
    let counts: Vec<(String, Json)> = onsets_v
        .iter()
        .filter(|(k, _)| k.starts_with("onset_"))
        .filter_map(|(k, v)| v.as_f64().map(|f| (k.clone(), Json::Int(f as i64))))
        .collect();
    Some(vec![
        ("dur".into(), Json::Num(onsets::round2(duration))),
        ("onsets".into(), Json::Obj(counts)),
    ])
}

#[cfg(test)]
mod tests {
    use super::*;

    /// `^\w{1,64}$` with re.ASCII — the id is the contract every route
    /// builds a path out of, so what it REFUSES is the load-bearing half.
    #[test]
    fn an_id_is_ascii_word_characters_and_at_most_sixty_four() {
        for ok in ["a", "_", "vigil", "storm_2", "TRACK9", "_src9"] {
            assert!(valid_id(ok), "{ok}");
        }
        assert!(valid_id(&"a".repeat(64)));
        assert!(!valid_id(&"a".repeat(65)));
        assert!(!valid_id(""));
        for no in [
            "a b", "a-b", "a.b", "a/b", "../etc", "a\0b", "a\n", ".hidden", "a%2f",
            // re.ASCII: Python's \w stops at the ASCII word set, so an
            // accented or non-Latin name is refused rather than reaching
            // the filesystem under a different spelling.
            "café", "трек", "🎃",
        ] {
            assert!(!valid_id(no), "{no}");
        }
        // 64 is a character count in Python; a multi-byte name that is
        // refused for its characters must not be accepted for its bytes.
        assert!(!valid_id(&"é".repeat(20)));
    }

    #[test]
    fn every_container_has_a_type_and_anything_else_is_a_download() {
        assert_eq!(mime("mp3"), "audio/mpeg");
        assert_eq!(mime("wav"), "audio/wav");
        assert_eq!(mime("flac"), "audio/flac");
        assert_eq!(mime("opus"), "audio/ogg");
        for ext in AUDIO_EXT {
            assert_ne!(mime(ext), "application/octet-stream", "{ext}");
        }
        assert_eq!(mime("MP3"), "application/octet-stream"); // never folded
        assert_eq!(mime(""), "application/octet-stream");
    }

    #[test]
    fn only_a_file_source_can_go_missing() {
        assert!(source_missing("file:/nowhere/_t_gone.wav"));
        assert!(!source_missing("file:/"));
        // A link is not a path: nothing about it is checkable on disk.
        assert!(!source_missing("https://example.invalid/x.mp3"));
        assert!(!source_missing(""));
        assert!(!source_missing("/nowhere/_t_gone.wav"));
    }
}
